### PR TITLE
[ONEM-24160] OCIContainer - Improve error reporting.

### PR DESCRIPTION
**Pull request by marcin-hajkowski-epam**

### CODE REVIEW

File: OCIContainer/OCIContainer.json [hunk 1]
```diff
@@ -39,6 +39,11 @@
             "properties": {
                 "success": {
                     "$ref": "#/definitions/success"
+                },
+                "error":{
+                        "summary": "An error message in case of a failure",
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.json:44]: For entire json API definition please use [schema](https://github.com/rdkcentral/Thunder/blob/master/Tools/JsonGenerator/schemas/interface.schema.json) defined in Thunder.

```json         
"errors": {
  "description": "Definition of possible error responses.",
  "type": "array",
  "items": {
    "description": "Definition of an error response.",
    "type": "object",
    "properties": {
      "description": {
        "description": "Description of the failure.",
        "type": "string"
      },
      "code": {
        "description": "Error code.",
        "type": "number"
      },
      "message": {
        "description": "Error message.",
        "type": "string"
      }
    },
    "required": [
      "code",
      "message"
    ]
  }
}
```
According to the schema errors (not error) param is an array of items with mandatory error code and message fields so this should be changed to:

```json
"errors": [
    {
        "description": "Internal Dobby Error",
        "$ref": "#/common/errors/internal_dobby_error"
    }
]
```
where "$ref": `#/common/errors/internal_dobby_error` can be define in common section of OCIContainer.json (before definitions) eg.

```json
"common": {
    "errors": {
        "internal_dobby_error": {
            "code": 1,
            "message": "ERROR_DOBBY_INTERNAL"
        },
        ...
    }
}
```
Please see example in Packager:
https://github.com/rdkcentral/rdkservices/blob/92b0ed86dda3d2ce845395f5ab61d1ef09bb3a1e/Packager/Packager.json

File: OCIContainer/OCIContainer.json [hunk 2]
```diff
@@ -49,6 +54,11 @@
             "summary": "Whether the request succeeded",
             "type": "boolean",
             "example": "true"
+        },
+        "error": {
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.json:58]: As above, follow schema and define error in common section.

File: OCIContainer/OCIContainer.json [hunk 3]
```diff
@@ -202,6 +212,11 @@
                     },
                     "success":{
                         "$ref": "#/definitions/success"
+                    },
+                    "error":{
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.json:216]: As above, follow schema and define error in common section

File: OCIContainer/OCIContainer.json [hunk 4]
```diff
@@ -243,6 +258,11 @@
...
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.json:262]: As above, follow schema and define error in common section

File: OCIContainer/OCIContainer.json [hunk 5]
```diff
@@ -422,6 +442,11 @@
...
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.json:446]: As above, follow schema and define error in common section

File: OCIContainer/doc/OCIContainerPlugin.md [hunk 6]
```diff
@@ -115,6 +115,7 @@ Executes a command inside a running container. The path to the executable must r
 | :-------- | :-------- | :-------- |
 | result | object |  |
 | result.success | boolean | Whether the request succeeded |
+| result.error | string | Error information |
```
Comment (piotr-serafin-epam) [inline OCIContainer/doc/OCIContainerPlugin.md:118]: Fix documentation to reflect Thunder schema meaning:

- errors array with items that consists mandatory error code and error message
- list all possible errors for this method

File: OCIContainer/doc/OCIContainerPlugin.md [hunk 7]
```diff
@@ -180,6 +181,7 @@ Gets information about a running container such as CPU, memory, and GPU usage (G
 | result.info.state | string | The container state (must be one of the following: *Invalid*, *Starting*, *Running*, *Stopped*, *Paused*) |
 | result.info.timestamp | integer | The timestamp for container information |
 | result.success | boolean | Whether the request succeeded |
+| result.error | string | Error information |
```
Comment (piotr-serafin-epam) [inline OCIContainer/doc/OCIContainerPlugin.md:184]: As above.

File: OCIContainer/doc/OCIContainerPlugin.md [hunk 8]
```diff
@@ -252,6 +254,7 @@ Gets the state of a currently running container.
 | result.containerId | string | The ID of a container as returned by `listContainers` |
 | result.state | string | The container state (must be one of the following: *Invalid*, *Starting*, *Running*, *Stopped*, *Paused*) |
 | result.success | boolean | Whether the request succeeded |
+| result.error | string | Error information |
```
Comment (piotr-serafin-epam) [inline OCIContainer/doc/OCIContainerPlugin.md:257]: As above.

File: OCIContainer/doc/OCIContainerPlugin.md [hunk 9]
```diff
@@ -350,6 +353,7 @@ Pauses a currently running container.
 | :-------- | :-------- | :-------- |
 | result | object |  |
 | result.success | boolean | Whether the request succeeded |
+| result.error | string | Error information |
```
Comment (piotr-serafin-epam) [inline OCIContainer/doc/OCIContainerPlugin.md:356]: As above.

File: OCIContainer/doc/OCIContainerPlugin.md [hunk 10]
```diff
@@ -396,6 +400,7 @@ Resumes a previously paused container.
...
```
Comment (piotr-serafin-epam) [inline OCIContainer/doc/OCIContainerPlugin.md:403]: As above.

File: OCIContainer/doc/OCIContainerPlugin.md [hunk 11]
```diff
@@ -554,6 +559,7 @@ Stops a currently running container.
...
```
Comment (piotr-serafin-epam) [inline OCIContainer/doc/OCIContainerPlugin.md:562]: As above.

File: OCIContainer/OCIContainer.cpp [hunk 12]
```diff
@@ -143,6 +143,7 @@ uint32_t OCIContainer::getContainerState(const JsonObject &parameters, JsonObjec
     int cd = GetContainerDescriptorFromId(id);
     if (cd < 0)
     {
+        response["error"] = "container descriptor can not be acquired";
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.cpp:146]: Please check Thunder controller implementation:
https://github.com/rdkcentral/Thunder/blob/master/Source/WPEFramework/Controller.cpp

eg 

```cpp
result->ErrorCode = Web::STATUS_NO_CONTENT;
result->Message = _T("Environment variable does not exist");
```

File: OCIContainer/OCIContainer.cpp [hunk 13]
```diff
@@ -199,6 +200,7 @@ uint32_t OCIContainer::getContainerInfo(const JsonObject &parameters, JsonObject
...
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.cpp:203]: As above.

File: OCIContainer/OCIContainer.cpp [hunk 14]
```diff
@@ -207,6 +209,7 @@ uint32_t OCIContainer::getContainerInfo(const JsonObject &parameters, JsonObject
     if (containerInfoString.empty())
     {
         LOGERR("Failed to get info for container %s", id.c_str());
+        response["error"] = "failed to get container info";
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.cpp:212]: As above.

File: OCIContainer/OCIContainer.cpp [hunk 15]
```diff
@@ -215,6 +218,7 @@ uint32_t OCIContainer::getContainerInfo(const JsonObject &parameters, JsonObject
     WPEC::OptionalType<WPEJ::Error> error;
     if (!WPEJ::IElement::FromString(containerInfoString, containerInfoJson, error)) {
         LOGERR("Failed to parse Dobby Spec JSON due to: %s", WPEJ::ErrorDisplayMessage(error).c_str());
+        response["error"] = "failed to parse dobby spec";
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.cpp:221]: As above.

File: OCIContainer/OCIContainer.cpp [hunk 16]
```diff
@@ -315,10 +319,10 @@ uint32_t OCIContainer::startContainerFromCryptedBundle(const JsonObject &paramet
                                        containerPath))
     {
         LOGERR("Failed to start container - sync mount request to omi failed.");
+        response["error"] = "mount failed";
```
Comment (piotr-serafin-epam) [inline OCIContainer/OCIContainer.cpp:322]: As above.

File: OCIContainer/OCIContainer.json (hunk 1 above)
Comment (marcin-hajkowski-epam) [reply OCIContainer/OCIContainer.json:44]: Hi @piotr-serafin-epam. As we discussed current proposition is compatible with other plugins error handling as well as with JSON schema which gives possibility to add optional items (eg error in result object).

Comment (piotr-serafin-epam) [reply OCIContainer/OCIContainer.json:44]: Ok sounds good, let's use result object instead to keep backward compatibility with existing OCIContainer clients. Please proceed with extending error reporting.

**Review by lukasz-wasylyk-red — APPROVED** [review]

**Pull request merged**